### PR TITLE
feat: Add strong types to `avm/res/sql/server` and it's children 

### PR DESCRIPTION
**Pull request by peterbud**
## Description

This PR is aiming to add strong types to parameters for `avm/res/sq/server` and it's children.

This PR contains typing for the `server`, `database` and `elastic-pool`. Other modules will be tackled later, as I'd like to see first that this is a good direction and approach adding these.

* Types are exported from children and imported into the main module
* To align with the latest API there is an SKU object for databases and elastic pools, instead of individual `skuName`, `skuTier` etc.
* Default values are provided in the child module, and the parent is not duplicating those default value assignment
* Also I have checked the latest API, and turned out that some of the new parameters were missing, so I went ahead and added them to have a complete coverage
* Updated `minCapacity` to enable fractional values both at database and elastic-pool modules
* As a consequence version number has been increased
* Updated tests where it was necessary. All tests are successfully completed

### Questions

* What is the best way to expose the types of the parameters of the child module with the least repetition? Currently I need to define a given module parameter for the module (with it's description), and then add the same to the type again with the same description. Is there a better  DRY way to do this?
* Also is there a better way to assign the values in the parent module to the parameters to the child module? Currently every parameter is assigned individually, maybe there is a better and shorter way to do that?

Contributes to #1141

## Pipeline Reference

| Pipeline |
| -------- |
|    [avm.res.sql.server](https://github.com/peterbud/bicep-registry-modules/actions/workflows/avm.res.sql.server.yml)      |

Pls note that in my subscription Maintenance Window feature is not available, hence the WAF test is failing

## Type of Change

- [ ] Update to CI Environment or utilities (Non-module affecting changes)
- [ ] Azure Verified Module updates:
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [x] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [x] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [x] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

**Comment by AlexanderSehr** [comment]
Hey @peterbud,
to your questions: @eriqua and I happened to discuss this today and will likely propose a spec to provide some guidelines. In general, this is what I'd suggest to do
- Start from the leafs of your module hierachy (i.e., the most nested child modules). There, you should have a parameter for each property of the implemented resource type, of which some may be objects & arrays of objects. For those you should implement a User Defined Type
- When you implement the User Defined Types, you should add an `@export()` annotation to make them available to calling modules. For now this is important for any of the parent modules, but eventually this will be a nice perk once we are able to published child modules too
- Now, going to the parents of those leaf-child modules: There you should have either an object or array of objects parameter for the child that is deployed (e.g., the blobServices module has an array (of objects) parameter for the containers to be deployed into the blobServies resource). As before, you should have a User Defined Type for those - but - while you should have the properties of the child being added to that user-defined-type, you should re-use the user defined types you added the `@export()` annotation for in the corresponding child. 
- If you do this in the tree all the way up, you will only need to repepat each child-modules parameters once - which should keep the types quite managable (size-wise).

There are a few smaller recommdantions we'd add to the above, such as
- ideally, a UDT should only describe how one single intance of an object looks like (even if it is for an array parameter). Whether a parameter is an array or nullable should be defined in the parameter, not the type (simply to make it less confusing). For example, rather have a parameter like `param roleAssignments roleAssignmentsType[]?` where the type is implemented like `type roleAssignmentType = {}` instead of `param roleAssignments roleAssignmentsType` where the type would be implemented like `type roleAssignmentType = {}[]?`
- Re-use types from the `types/avm-common-types` module (we just published) for the AVM specs to reduce the number of UDTs in your module - be it only for the AVM specs. But hey, it's something 😄 

Now to your question regarding 'passing each property/parameter down individually'. I have a strong guess you're referring to examples like
https://github.com/Azure/bicep-registry-modules/blob/644e64f6cf91040cf594ded1cb67c2d50d4e36e0/avm/res/sql/server/main.bicep#L248-L259

The simple answer is - no 😄 There is no 'easier' way by default. Bicep always wants you to declare the handover explicitely. However, you can shorten it slightly, which brings me to another piece of advice:
- If you pass values down to any referenced module (child or otherwise) you should check of those parameters have a default in the referenced template. In 9 out of 10 cases, the default in the parent (behind the `??`) will be a duplicate of the default in the child. In the above example that would, for example, be the case for the `skuTier` parameter.
- What you should do, is remove those defaults from the parent. In this case to change `skuTier: database.?skuTier ?? 'GeneralPurpose'` to `skuTier: database.?skuTier`. Why? Well because since nullable types were introduced, Bicep will automatically assume a referenced module's defaut values if you pass null into them. So those explicit defaults in the calling module are not needed anymore (unless you want to specify something else that the default in the referenced module, of course)
- Back when we started, there was no other way then to duplicate those values. Now, we just did not get to 'clean' all the modules up
- This won't save you from specifying the properties individually, but at least it will be a simple pass through

I hope to above provided some guidance.
@eriqua, feel free to add to it if something comes to mind.

File: avm/res/sql/server/elastic-pool/main.bicep
Comment (AlexanderSehr) [inline avm/res/sql/server/elastic-pool/main.bicep:8]: Please note, this parameter is only required if using this module in a standalone deployment. If you use it via its parent, the value would actually be optional, as the `server/main.bicep` module would pass its name through. Hence we specified those parameters (that is, parent resource names) as 'Conditional'. With UDT's we could actually change that - but I'd suggest to do this 'systematically' for all modules instead of just individual ones.

File: avm/res/sql/server/elastic-pool/main.bicep
Comment (AlexanderSehr) [inline avm/res/sql/server/elastic-pool/main.bicep:151]: Defined as optional

File: avm/res/sql/server/elastic-pool/main.bicep
Comment (AlexanderSehr) [inline avm/res/sql/server/elastic-pool/main.bicep:156]: I'd recommend to move this type (including the export, don't remove that 😄) into the parent module

File: avm/res/sql/server/database/main.bicep
Comment (AlexanderSehr) [inline avm/res/sql/server/database/main.bicep:45]:
```suggestion
param elasticPoolResourceId string?
```

as per [spec](https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-snfr22---category-inputs---parametersvariables-for-resource-ids).

File: avm/res/sql/server/database/main.bicep
Comment (AlexanderSehr) [inline avm/res/sql/server/database/main.bicep:83]:
```suggestion
param minCapacity int = 0
```
as per [spec](https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-snfr14---category-inputs---data-types). We both know the Resource Provider won't like that, so the casting must happen where the value is used (e.g., `'${minCapacity}'`)

File: avm/res/sql/server/database/main.bicep
Comment (AlexanderSehr) [inline avm/res/sql/server/database/main.bicep:95]: Is that a resource ID? If so, it should contain `resourceId` in the name

File: avm/res/sql/server/database/main.bicep
Comment (AlexanderSehr) [inline avm/res/sql/server/database/main.bicep:98]: same as previous comments about resource Ids (if applicable)

File: avm/res/sql/server/database/main.bicep
Comment (AlexanderSehr) [inline avm/res/sql/server/database/main.bicep:119]: same comment as with the other resource Ids

File: avm/res/sql/server/database/main.bicep
Comment (AlexanderSehr) [inline avm/res/sql/server/database/main.bicep:122]: and again - same comment as with the other resource IDs 😄

File: avm/res/sql/server/database/main.bicep
Comment (AlexanderSehr) [inline avm/res/sql/server/database/main.bicep:128]: I don't thing it had `true` default by accident. Was likely implemented this way to be WAF aligned 😉

**Review by AlexanderSehr — CHANGES_REQUESTED** [review]
Hey @peterbud, I now added quite a few comments that will apply to all files you touched. I'll hence stop here for a moment and wait for you to respond to the current comments :)

File: avm/res/sql/server/database/main.bicep
Comment (peterbud) [reply avm/res/sql/server/database/main.bicep:128]: I'd argue that default zone redundancy should be off for the following reasons:

* At the Azure Portal zone redundancy off is the default setting (you can check at the Azure portal when creating a database)
* Zone redundancy is **only** available at [selected regions](https://microsoft.sharepoint.com/:f:/r/teams/csscm/Shared%20Documents/CapacityManagementTool/[KB](https://microsoft.sharepoint.com/:f:/r/teams/csscm/Shared%20Documents/CapacityManagementTool/KB?csf=1&web=1&e=SbrEU9)?csf=1&web=1&e=SbrEU9)
* Also zone redundancy is not a security feature but a _high availability_ feature.

I agree that for a WAF deployment zone redundancy is necessary, but for "plain" deployments turning it on be by default is leading to unnecessary deployment issues (not available everywhere).

**Comment by peterbud** [comment]
@AlexanderSehr Thanks for the quick review, I'll make the appropriate changes.

File: avm/res/sql/server/database/main.bicep
Comment (peterbud) [reply avm/res/sql/server/database/main.bicep:83]: `int` would not work. If you define as `int`, you cannot pass in fractional values, like 0.5 or 0.25, which is needed for this param.

Even the [API definition](https://learn.microsoft.com/en-us/azure/templates/microsoft.sql/servers/databases?pivots=deployment-language-bicep) is saying: `minCapacity: json('decimal-as-string')`

**Comment by peterbud** [comment]
I have marked the items as resolved, where I could resolve them, so PR ready for further review

**Comment by peterbud** [comment]
I have also added the types for backup sub-modules, with them at least the current database and elastic-pool modules are complete

File: avm/res/sql/server/database/main.bicep
Comment (AlexanderSehr) [reply avm/res/sql/server/database/main.bicep:83]: Hey @peterbud, thanks for elaburating. Then that's my bad - I was under the impression it is a full number, not a decimal. Resolving

**Comment by AlexanderSehr** [comment]
> > * If you do this in the tree all the way up, you will only need to repepat each child-modules parameters once - which should keep the types quite managable (size-wise).
> 
> I'm not sure I follow you here. Let's take an example: a child module called "child" and a "parent" module.
> 
> If a child has a parameter called "age", I define it in the child module, like:
> 
> ```bicep
> @description('Required. The age of the child.')
> param age int
> ```
> 
> So far so good.
> 
> But if in the parent I have a collection of children, I need to define a UDT, if I understand your suggestion. However I need to define here **again** the same property:
> 
> ```bicep
> ...
> @description ('Optional. The children of the parent')
> param children: childrenPropType[]?
> 
> @export()
> @description('The children type')
> type childrenPropType = {
>   @description('Required. The age of the child.')
>   age: int
> 
> ... //other props
> }
> ```
> 
> So this is the "repeating only once" solution, right? I was wondering how to avoid defining the same stuff twice, but as you describe its not feasible right now?

Hey @peterbud,
that's correct. Please excuse the presumably confusing explanation. Without 'showing' it's not an easy concept to convey 😄 

Under consideration of Bicep's current features, we have to do it like you describe and repeat the property once. What I was try to get to was to say that 'at least' the parent of the second module you show above does not need to describe `age: int` yet again, as it would instead import the `childPropType` type.

**Comment by peterbud** [comment]
Thanks, you now it's clear!

**Review by jikuja — COMMENTED** [review]
Really nice additions. This makes usage of the module much more easier. Few comments on review. Nothing serious.

File: avm/res/sql/server/database/backup-long-term-retention-policy/main.bicep
Comment (jikuja) [reply avm/res/sql/server/database/backup-long-term-retention-policy/main.bicep:18]: Missed the correct lines:
```
@description('Optional. Weekly retention in ISO 8601 duration format.')
param monthlyRetention string?

@description('Optional. Monthly retention in ISO 8601 duration format.')
param weeklyRetention string?
```

Looks like descriptions should be swapped.

File: avm/res/sql/server/audit-settings/main.bicep
Comment (peterbud) [reply avm/res/sql/server/audit-settings/main.bicep:11]: Indeed, thank you!

File: avm/res/sql/server/database/backup-long-term-retention-policy/main.bicep
Comment (peterbud) [reply avm/res/sql/server/database/backup-long-term-retention-policy/main.bicep:37]: Because the latest API version has an issue, and does not recognize `backupStorageAccessTier` and `makeBackupsImmutable` properties
See the [2024-08-01 version](https://github.com/Azure/bicep-types-az/blob/main/generated/sql/microsoft.sql/2023-08-01-preview/types.md#properties-133) (wrong)
Vs the [2024-05-01 version](https://github.com/Azure/bicep-types-az/blob/main/generated/sql/microsoft.sql/2023-05-01-preview/types.md#longtermretentionpolicyproperties) (which is good)

Comment (jikuja) [reply avm/res/sql/server/database/backup-long-term-retention-policy/main.bicep:37]: Good to know. 

For this kind of implementation decision it could be a good idea to write ADR as a comment.

File: avm/res/sql/server/database/main.bicep
Comment (peterbud) [reply avm/res/sql/server/database/main.bicep:20]: Good catch, probably -1 would be a better default. I'll update

File: avm/res/sql/server/database/main.bicep
Comment (AlexanderSehr) [reply avm/res/sql/server/database/main.bicep:128]: Hey @peterbud,
I get your point, but the so-called 'reliability' pillar of WAF is one of the enforced ones for AVM. I'm pretty sure sooner or later there'll also be a failing PSRule test if we don't enable the setting. For this reason we usually aways enabled the setting for AVM's resource modules.
Lastly, the Portal is (unfortunately) not a reliable source for settings.

Anyways, let me loop @jtracey93 & @eriqua in here to collect their thoughts as well.

File: avm/res/sql/server/audit-settings/main.bicep
Comment (AlexanderSehr) [inline avm/res/sql/server/audit-settings/main.bicep:19]:
```suggestion
param auditActionsAndGroups string[] = [
```

File: avm/res/sql/server/elastic-pool/main.bicep
Comment (AlexanderSehr) [inline avm/res/sql/server/elastic-pool/main.bicep:46]: Leftover?

File: avm/res/sql/server/tests/e2e/elasticPool/main.test.bicep
Comment (AlexanderSehr) [inline avm/res/sql/server/tests/e2e/elasticPool/main.test.bicep:18]:
```suggestion
param serviceShort string = 'ssep'
```

File: avm/res/sql/server/database/main.bicep
Comment (eriqua) [reply avm/res/sql/server/database/main.bicep:128]: Hi @peterbud, echoing what @AlexanderSehr mentioned above.

You can reference this FAQ page to know more about [What AVM means by WAF aligned](https://azure.github.io/Azure-Verified-Modules/faq/#what-does-avm-mean-by-waf-aligned)
In particular, the further note mentioning how both reliability and security are prioritized.

Let me add that your point about zone redundancy not being available for all locations is totally valid. 

However, as mentioned by the same note, defaulted values can be anyway altered via input parameters if required. 

This means zone redundancy should be explicitly disabled if using a non supporting location. Generally, the intention is to make it more difficult rather than easier to deploy a non WAF aligned resource.

@jtracey93 happy to hear your thoughts as well

Comment (peterbud) [reply avm/res/sql/server/database/main.bicep:128]: OK, no worries, then let's keep `zoneRedundant` by default on.

File: avm/res/sql/server/audit-settings/main.bicep
Comment (peterbud) [reply avm/res/sql/server/audit-settings/main.bicep:19]: Good idea

File: avm/res/sql/server/elastic-pool/main.bicep
Comment (peterbud) [reply avm/res/sql/server/elastic-pool/main.bicep:46]: I was not sure why the description specifically mentions the serverless pool. I had the impression that for elastic pool you need a provisioned server. Anyhow, let's figure out later.

File: avm/res/sql/server/main.bicep
Comment (AlexanderSehr) [inline avm/res/sql/server/main.bicep:299]: Leftover

File: avm/res/sql/server/main.bicep
Comment (AlexanderSehr) [inline avm/res/sql/server/main.bicep:707]: Optional/Required?

**Review by AlexanderSehr — COMMENTED** [review]
Nothing more to add from my side :) @emichellecarter, do you want to give the PR a final look as the co-owner?

**Comment by jikuja** [comment]
I was wondering new default settings for avm/res/sql/server/audit-settings/main.bicep:

* Is there any performance degraration when `param state string = 'Enabled'` but logging targets are disabled
* Is there any perfromance degraration when `param state string = 'Enabled'` and `isAzureMonitorTargetEnabled bool = true` but no diagnostics settings are defined
* Should avm/res/sql/server/main.bicep or avm/res/sql/server/audit-settings/main.bicep  somewhere mention that auditing module settings does not turn on any log persist methods?

I personally would like to see server module to deploy `Microsoft.Sql/servers/auditingSettings` resource with good default settings to maintain sane state of the resource:

* if user does not supply `auditSettings` => deploy resource with `state = 'Disabled'` (Here is a good set of discussions of logging should be turned on and how much configuration is needed for WAF. At the moment logging requires extra config and turning on logging without configuring any persistancy IMO gives false sense of security)
* if user removes `auditSettings` => deploy resource with `state = 'Disabled'`
* if user adds `auditSettings` => deploy resource with users' config

----
In  one of the earlier tickets there was short mention of the database logging configs. That is a good to keep in mind when re-writing server level defaults and server level logging types/documentation.

File: avm/res/sql/server/main.bicep
Comment (jikuja) [reply avm/res/sql/server/main.bicep:707]: IMO required. It user makes conscious choise to configure audit logging then they should also select if logging is enabled or disabled.

File: avm/res/sql/server/main.bicep
Comment (peterbud) [reply avm/res/sql/server/main.bicep:299]: Actually I forgot to uncomment them after defining the objects. Good catch.

File: avm/res/sql/server/main.bicep
Comment (peterbud) [reply avm/res/sql/server/main.bicep:707]: Optional, default is 'Enabled'

**Comment by peterbud** [comment]
Hi @jikuja thanks for the feedback

> I was wondering new default settings for avm/res/sql/server/audit-settings/main.bicep:
> 
> * Is there any performance degraration when `param state string = 'Enabled'` but logging targets are disabled

I'm not aware of any performance degradation. If you have specific know case I'd love to hear that.

> * Is there any perfromance degraration when `param state string = 'Enabled'` and `isAzureMonitorTargetEnabled bool = true` but no diagnostics settings are defined

Same as above: I'm not aware of any performance degradation

> I personally would like to see server module to deploy `Microsoft.Sql/servers/auditingSettings` resource with good default settings to maintain sane state of the resource:
> 
> * if user does not supply `auditSettings` => deploy resource with `state = 'Disabled'` (Here is a good set of discussions of logging should be turned on and how much configuration is needed for WAF. At the moment logging requires extra config and turning on logging without configuring any persistancy IMO gives false sense of security)
> * if user removes `auditSettings` => deploy resource with `state = 'Disabled'`
> * if user adds `auditSettings` => deploy resource with users' config
> 
> In one of the earlier tickets there was short mention of the database logging configs. That is a good to keep in mind when re-writing server level defaults and server level logging types/documentation.

Auditing should be turned ON by default with the recommended settings we have in this PR. Here are the reasons:

* Look at the current pipeline runs: pre-flight failed rules (before this PR): Auditing rule is failing as no audit settings was configured by default,
* This Azure rule is pretty clear that at least a basic auditing should be in place by default: https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.SQL.Auditing/ - and that is what this PR is putting into practice

In practice this means:
* If user does not specify any `auditingSettings`, the default will deploy a basic audit settings as the above mentioned rule requires (and pre-flight will e successful): 
* If the user specifies `auditingSettings`: {state: 'Disabled'} auditing can be turned off
* If the user specifies an `auditingSettings` with Log persistence etc. that will be obviously honored.

**Comment by jikuja** [comment]
The work with parameter typing is really important and missing typing was one the biggest issue when I used module. Good work!


> Auditing should be turned ON by default with the recommended settings we have in this PR. Here are the reasons:
> 
> * Look at the current pipeline runs: pre-flight failed rules (before this PR): Auditing rule is failing as no audit settings was configured by default,
> * This Azure rule is pretty clear that at least a basic auditing should be in place by default: https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.SQL.Auditing/ - and that is what this PR is putting into practice

**Saying this first: examples are easy to add afterwards. Things I write should not cause any delays with the merge.**

I personally feel that template or documentation should do more than just satisfy PsRule: at least there should be documentation how to enable logging with diagnosticsSetting. 

I till remember how hard it was to do first IaC template for Azure SQL because the only documentation for proper diagnostics setting creation was available on on quickstarts and/or by checking what portal does when enabling  auditiong.

Because Azure SQL has more settings on two levels both should be handled by template  or by the documentation to make module usage easier. It is really good to see that if `allLogs` is used by default if categories are not supplied by the user.

**Comment by AlexanderSehr** [comment]
> The work with parameter typing is really important and missing typing was one the biggest issue when I used module. Good work!
> 
> > Auditing should be turned ON by default with the recommended settings we have in this PR. Here are the reasons:
> > 
> > * Look at the current pipeline runs: pre-flight failed rules (before this PR): Auditing rule is failing as no audit settings was configured by default,
> > * This Azure rule is pretty clear that at least a basic auditing should be in place by default: https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.SQL.Auditing/ - and that is what this PR is putting into practice
> 
> **Saying this first: examples are easy to add afterwards. Things I write should not cause any delays with the merge.**
> 
> I personally feel that template or documentation should do more than just satisfy PsRule: at least there should be documentation how to enable logging with diagnosticsSetting.
> 
> I till remember how hard it was to do first IaC template for Azure SQL because the only documentation for proper diagnostics setting creation was available on on quickstarts and/or by checking what portal does when enabling auditiong.
> 
> Because Azure SQL has more settings on two levels both should be handled by template or by the documentation to make module usage easier. It is really good to see that if `allLogs` is used by default if categories are not supplied by the user.

Hey @jikuja,
absolutely fair points. Ideally, there should be a few places that can help
- There should be a User-Defined-Type for every object / object-array with the parameters & metadata to explain how to configure xzy setting
- The `Usage Examples` in the `readme.md` file pull the examples from the test cases. I.e., if e.g., a [max test](https://github.com/Azure/bicep-registry-modules/tree/main/avm/res/sql/server#example-5-using-large-parameter-set) implement diagnostic settings (to test the feature), the readme should show the same. In an ideal case, this is also mentioned in the test's metadata description, which is also pulled in. 
One could of course argue there should also be a code example that just shows the diagnostic settings (or any other feature for that matter). The challenge with those is usually that one has to maintain them. In the worst case you end up with outdated/stale code examples if somebody forgets to update them with every change. Hence we chose the approach of pulled the examples from the deployment tests.

**Comment by peterbud** [comment]
> * The `Usage Examples` in the `readme.md` file pull the examples from the test cases. I.e., if e.g., a [max test](https://github.com/Azure/bicep-registry-modules/tree/main/avm/res/sql/server#example-5-using-large-parameter-set) implement diagnostic settings (to test the feature), the readme should show the same. In an ideal case, this is also mentioned in the test's metadata description, which is also pulled in.

We can add later more test files / examples to address different use cases, admittedly there is a high number of different setups which might be interesting for the users.

**Comment by AlexanderSehr** [comment]
> > * The `Usage Examples` in the `readme.md` file pull the examples from the test cases. I.e., if e.g., a [max test](https://github.com/Azure/bicep-registry-modules/tree/main/avm/res/sql/server#example-5-using-large-parameter-set) implement diagnostic settings (to test the feature), the readme should show the same. In an ideal case, this is also mentioned in the test's metadata description, which is also pulled in.
> 
> We can add later more test files / examples to address different use cases, admittedly there is a high number of different setups which might be interesting for the users.

True. I guess the one thing to always wheigh against is that each test means another deployment which requires capacity & agents. For most modules it's not an issue, while for some there are quite specific restrictions

**Comment by peterbud** [comment]
> True. I guess the one thing to always wheigh against is that each test means another deployment which requires capacity & agents. For most modules it's not an issue, while for some there are quite specific restrictions

Maybe it would be a nice feature to be able to add more examples to the generated README.md without adding more test cases. I'm not sure how difficult that would be though, as never looked at the README generation.

**Comment by peterbud** [comment]
Is there anything outstanding to this PR, or could it be merged?

**Comment by AlexanderSehr** [comment]
> > True. I guess the one thing to always wheigh against is that each test means another deployment which requires capacity & agents. For most modules it's not an issue, while for some there are quite specific restrictions
> 
> Maybe it would be a nice feature to be able to add more examples to the generated README.md without adding more test cases. I'm not sure how difficult that would be though, as never looked at the README generation.

It would be doable. But not advisable as this means we'd have a lot of code that isn't tested and may turn stale in no time. Some sort of validation would always be preferred. :)

**Comment by AlexanderSehr** [comment]
> Is there anything outstanding to this PR, or could it be merged?

I don't think there's anything more from my side. Ideally @emichellecarter would give it her blessing as the co-owner. But if she is not available, the repo maintainers can help out.

**Comment by peterbud** [comment]
AFAIK @emichellecarter does not have a bandwidth ATM, so I don't think she would be able to review this PR

**Review by AlexanderSehr — DISMISSED** [review]

**Comment by AlexanderSehr** [comment]
> AFAIK @emichellecarter does not have a bandwidth ATM, so I don't think she would be able to review this PR

Gotcha @peterbud, thank you.
Please note that there is a conflict with the `main.json`. Please resolve it and (**important**) run the `Set-AVMModule` utility again after to make sure the template hash etc. is really up to date after the merge.

**Comment by peterbud** [comment]
> Please note that there is a conflict with the `main.json`. Please resolve it and (**important**) run the `Set-AVMModule` utility again after to make sure the template hash etc. is really up to date after the merge.

Done.

**Review by AlexanderSehr — APPROVED** [review]
Thank you @peterbud,
off we go and fingers crossed. I'll keep an eye on the workflow.

**Comment by peterbud** [comment]
Looks like a new static validation test failed. I'll look into that

**Comment by peterbud** [comment]
I have seen that you went ahead and created the fix. Thx

**Comment by AlexanderSehr** [comment]
That's right. The PR is ready to go 💪

**Pull request merged**
